planner/core/indexmerge_path.go

```diff
@@ -87,26 +111,37 @@ func (ds *DataSource) generateIndexMergePath() error {
 		}
 	}
 
-	if isPossibleIdxMerge && sessionAndStmtPermission && needConsiderIndexMerge && ds.tableInfo.TempTableType != model.TempTableLocal {
-		err := ds.generateAndPruneIndexMergePath(indexMergeConds, ds.indexMergeHints != nil)
-		if err != nil {
-			return err
-		}
-	} else if len(ds.indexMergeHints) > 0 {
+	if !needConsiderIndexMerge {
+		warningMsg = "IndexMerge is inapplicable or disabled. "
+		return nil
+	}
+	regularPathCount := len(ds.possibleAccessPaths)
+	if err := ds.generateAndPruneIndexMergePath(indexMergeConds); err != nil {
+		return err
+	}
+
+	// If without hints, it means that `enableIndexMerge` is true
+	if len(ds.indexMergeHints) == 0 {
+		return nil
+	}
+	// If len(indexMergeHints) > 0, then add warnings if index-merge hints cannot work.
+	if regularPathCount == len(ds.possibleAccessPaths) {
 		ds.indexMergeHints = nil
-		var msg string
-		if !isPossibleIdxMerge {
-			msg = "No available filter or available index."
-		} else if !sessionAndStmtPermission {
-			msg = "Got no_index_merge hint or tidb_enable_index_merge is off."
-		} else if ds.tableInfo.TempTableType == model.TempTableLocal {
-			msg = "Cannot use IndexMerge on temporary table."
-		}
-		msg = fmt.Sprintf("IndexMerge is inapplicable or disabled. %s", msg)
-		stmtCtx.AppendWarning(errors.Errorf(msg))
-		logutil.BgLogger().Debug(msg)
+		ds.ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("IndexMerge is inapplicable"))
+		return nil
 	}
 
+	// If len(indexMergeHints) > 0 and some index-merge paths were added, then prune all other non-index-merge paths.
+	ds.possibleAccessPaths = ds.possibleAccessPaths[regularPathCount:]
+	minRowCount := ds.possibleAccessPaths[0].CountAfterAccess
+	for _, path := range ds.possibleAccessPaths {
+		if minRowCount < path.CountAfterAccess {
+			minRowCount = path.CountAfterAccess
```

Comment:
It seems that `minRowCount` is the maximum of each path's `CountAfterAccess`, which is confusing. Since the PR is a refactor, we can solve the problem in the future PR.